Provide utils for font metadata

lionel- opened the pull request:
@davidgohel could you try this branch and see if it fills your needs please.

lionel- committed:
Provide utils for font metadata

See #5

lionel- commented on R/utils.R:
```diff
@@ -55,3 +92,89 @@ str_standardise <- function(name) {
   sub("\\..+$", "", path)
 }
+
+gather_tree <- function(x, depth) {
```
@hadley just wrote this `gather_tree()` util that may be useful in purrr or tidyr.

It takes a tree of data frames and a depth and gather the data frames together. `depth` can be a character vector whose elements are passed to `.id` in `bind_rows()`.

It's accompanied with `spread_attributes()` which creates a data frame out of an atomic vector and its attributes (which should be as long as the vector).

There is also `at_bottom()` which I use to map `spread_attributes()` on a tree of vectors + metadata to create data frames at the bottom. I think that's altogether a nice little idiom.

hadley replied:
I could've sworn I wrote a reply to this but maybe it got lost?

I haven't thought much about this sort of data structure, but it does have some parallels with subplots in ggvis. The idea is to be able to have a hierarchical arrangement of plots (e.g. barcharts of rainfall for each city, overlaid on a map). I was thinking about having a tree of plot data where the top level would (e.g.) give the position, size, background colour of the sub plot, and the next level would give the data of the barchart.

But I think that requires a slightly more general structure to what you're using here because the nodes can also have data, not just the children.

lionel- replied:
Trees also have applications when dealing with stuff like json, but I haven't worked a lot with this kind of thing either.

davidgohel commented:
Thanks, I just had a look, it looks like what I need.

lionel- committed:
Document new font metadata utils

hadley commented on R/font-bitstream-vera.R:
```diff
@@ -34,21 +35,21 @@ font_bitstream_vera <- function(variant = "sans", style = "roman",
-    `bold-oblique` = "VeraBI"
+font_bitstream_vera_files <- font_files(font_bitstream_vera,
+  sans = list(
+    roman = font_sans("Vera"),
```
I think using "plain" would be more inline with standard R terminology

lionel- replied:
hmm when I look in `?par` they mention "sans", "serif" and "mono" for `family` and "plain", "bold", "italic" and "bold italic" for `font` (which in fontquiver is called `face`).

hadley replied:
Sorry, I mean plain instead of "roman" (and you use "book" below)

lionel- replied:
ah yes but these are the fontconfig categories. The main data structure uses the fontconfig nomenclature since it is a superset of the R nomenclature. Then the R categories are recorded by the font constructors `font_sans(bold = FALSE, italic = FALSE)`, `font_serif(bold = FALSE, italic = FALSE)`, etc.

This way the regular getter provide access to all versions of a font, and you can use the new helpers provided in this PR (`font_info()`, `font_info_files()`) to request a font using the R categories.

Is this approach reasonable?

hadley replied:
Ah got it, that makes sense.

hadley commented on R/utils.R:
```diff
@@ -55,3 +92,89 @@ str_standardise <- function(name) {
+    call(fun, lhs, rhs)
+  }
+  pattern <- Reduce(bind_call, fs[-1], fs[[1]])
+  args <- list(df, pattern, substitute())
```
I think you need `args$drop <- FALSE` here to be safe (and below in `filter_first()` you also need `drop = FALSE`)

lionel- committed:
Handle weights and face = "symbol" in metadata

lionel- committed:
Prevent `[` from dropping dimensions

lionel- committed:
Provide font_variants() and font_families()

Also font_variant() and font_family() for a single style or face.

These will be useful to provide a set of fonts to other tools such as
svglite or htmlFontDependency.

lionel- committed:
Add all metadata in font_file objects

lionel- committed:
Remove `ext` arguments from UI

And provide all files in `font_file` objects

lionel- committed:
Make collections of fonts S3 objects

lionel- committed:
Use font_faces and font_styles names for subcollections

lionel- committed:
Improve UI and CSS generation of htmlFontDependency()

- Can accept any font or collection of fonts
- Now uses canonical font name
- Writes font-style and font-weight CSS attributes

Example of generated CSS file:

  #> htmlFontDependency(font_families("Bitstream Vera"))

  @font-face {
    font-family: 'Bitstream Vera Sans';
    font-style: normal;
    font-weight: 400;
    src:
      url('Vera.woff') format('woff'),
      url('Vera.ttf')  format('truetype');
  }

  @font-face {
    font-family: 'Bitstream Vera Sans';
    font-style: italic;
    font-weight: 400;
    src:
      url('VeraIt.woff') format('woff'),
      url('VeraIt.ttf')  format('truetype');
  }

  @font-face {
    font-family: 'Bitstream Vera Sans';
    font-style: normal;
    font-weight: 700;
    src:
      url('VeraBd.woff') format('woff'),
      url('VeraBd.ttf')  format('truetype');
  }

  @font-face {
    font-family: 'Bitstream Vera Sans';
    font-style: italic;
    font-weight: 700;
    src:
      url('VeraBI.woff') format('woff'),
      url('VeraBI.ttf')  format('truetype');
  }

  @font-face {
    font-family: 'Bitstream Vera Serif';
    font-style: normal;
    font-weight: 400;
    src:
      url('VeraSe.woff') format('woff'),
      url('VeraSe.ttf')  format('truetype');
  }

  ...

lionel- commented:
Last commit should make it real easy to use any font with `htmlFontDependency()` /  `htmlSvg()`. Is this still missing something @davidgohel? (cf commIt message)

lionel- committed:
Prettify names returned by font_variants()

lionel- committed:
Export splice_fonts()

lionel- committed:
Add fullname to font_file attributes

lionel- committed:
Remove getter attribute in font data

lionel- committed:
Better names

lionel- committed:
Add and export fontset_list()

To get list of available fonts

lionel- committed:
Reorganise font getters

lionel- committed:
Unexport and undocument font_bitstream_vera() and font_dejavu()

And use fontset_ prefix

lionel- committed:
Rename font_info() to fontset_info()

lionel- committed:
Move fontset helpers to fontset.R

lionel- committed:
Move fontset specs to fontset- prefixed files

And move font-info.R to font-getters.R

lionel- committed:
Fix documentation

lionel- committed:
Use fontset_props() in fontset_info()

lionel- committed:
Add support for Liberation fonts

lionel- committed:
Include Symbola font in fontquiver

lionel- committed:
Rename symbol_font() to font_symbol()

And rename all constructors to make room for font_symbol()

lionel- committed:
Remove fontLiberation from remotes

lionel- committed:
Update NEWS file

lionel- committed:
Bump version and some more NEWS

hadley commented on DESCRIPTION:
```diff
@@ -1,18 +1,31 @@
     testthat,
     htmltools
-License: GPL-3
+License: file LICENSE
```
I think this should be `GPL-3 | file LICENSE` ?

lionel- replied:
What is the meaning of `|` for the License field? I can't find this information in R exts

hadley replied:
It means or. `+` might be more appropriate, I can never remember.

hadley commented on R/font-getters.R:
```diff
@@ -0,0 +1,136 @@
+#' Get elements of a fontset
+#'
+#' \code{fonts()} returns the list of all \code{font_file} objects for
+#' a given fontset. \code{font()} returns one specific font.
```
What is a fontset?

lionel- replied:
This is a collection of font families (e.g. Bitstream Vera or Liberation), I'll add a note.

hadley commented on R/font-getters.R:
```diff
@@ -0,0 +1,136 @@
+#' a given fontset. \code{font()} returns one specific font.
+#' \code{font_symbol()} extracts the font marked as symbol font.
+#' @inheritParams fontset_info
+#' @param variant Font variant, as per Fontconfig's nomenclature.
```
fontconfig? Might be useful to include brief descriptions of the nomenclature to save the user some googling. A couple of examples might be sufficient.

hadley commented on R/font-getters.R:
```diff
@@ -0,0 +1,136 @@
+
+#' @rdname font_families
+#' @export
+font_faces <- function(fontset, family) {
```
Use `family = c("sans", "serif", ...)` and `match.arg()` here?

hadley commented on R/html-dependency.R:
```diff
@@ -1,66 +1,81 @@
     "@font-face {",
-    sprintf("  font-family: '%s';", id),
-    "  src: ",
-    "    url('%s.woff') format('woff'),                    /* Modern Browsers */",
```
I think it would be useful to preserve these comments

lionel- replied:
those comments are now outdated: http://caniuse.com/#feat=woff

hadley replied:
In that case, should we drop support for ttf? It doesn't seem like it's any better supported

hadley commented on R/fontset-dejavu.R:
```diff
@@ -0,0 +1,54 @@
+#' @references \url{http://dejavu-fonts.org}
+#' @include fontset.R
+fontset_dejavu <- function(variant = "sans", style = "book") {
+  if (system.file("fonts", package = "fontHeavy") == "") {
```
Should this now be be `fontDejaVu`?

lionel- replied:
this is the Github package that holds fonts too heavy for CRAN, so a generic name is probably better since it can potentially hold multiple fonts.

hadley replied:
At some point it'll get too heavy even for github? One font per package seems like a simpler and more generalisable system.

lionel- committed:
Use match.arg() in font_faces()

lionel- committed:
State what a fontset is

lionel- committed:
Add helper functions to get information about variants and styles

lionel- committed:
Rename fontHeavy to fontDejaVu

lionel- committed:
Prettify font names in error messages

lionel- committed:
Drop support for ttf in htmlFontDependency()

lionel- committed:
Change s3 objects font_file to font

Makes sense that a fontset would contain `font` objects

lionel- committed:
Update LICENSE field

hadley commented:
LGTM
